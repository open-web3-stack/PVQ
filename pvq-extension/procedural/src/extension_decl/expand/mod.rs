mod extension;
mod helper;

use crate::extension_decl::Def;
use proc_macro2::TokenStream as TokenStream2;
use quote::ToTokens;

pub fn expand(mut def: Def) -> TokenStream2 {
    let extension_expanded = extension::expand_extension(&mut def);

    let new_items = quote::quote! {
        #extension_expanded
    };

    def.item
        .content
        .as_mut()
        .expect("This is checked by parsing")
        .1
        .push(syn::Item::Verbatim(new_items));
    def.item.into_token_stream()
}
