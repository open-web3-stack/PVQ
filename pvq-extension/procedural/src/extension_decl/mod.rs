mod expand;
pub(crate) mod parse;
pub use parse::Def;
use proc_macro::TokenStream;
use proc_macro2::TokenStream as TokenStream2;
use syn::spanned::Spanned;

mod keyword {
    syn::custom_keyword!(dev_mode);
}

pub fn extension_decl(attr: TokenStream, item: TokenStream) -> TokenStream {
    if !attr.is_empty() {
        let msg = "Invalid #[extension_decl] macro call: unexpected attribute. Macro call must be bare, such as `#[extension_decl]`.";
        let span = TokenStream2::from(attr).span();
        return syn::Error::new(span, msg).to_compile_error().into();
    }

    let item = syn::parse_macro_input!(item as syn::ItemMod);
    match parse::Def::try_from(item) {
        Ok(def) => expand::expand(def).into(),
        Err(e) => e.to_compile_error().into(),
    }
}
