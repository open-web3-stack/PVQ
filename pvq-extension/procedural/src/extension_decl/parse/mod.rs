pub mod extension;
mod helper;

use crate::utils::generate_crate_access;
use syn::spanned::Spanned;

mod keyword {
    syn::custom_keyword!(extension_decl);
    syn::custom_keyword!(extension);
}
pub struct Def {
    pub item: syn::ItemMod,
    pub extension: extension::Extension,
    pub pvq_extension: syn::Path,
}

impl Def {
    pub fn try_from(mut item: syn::ItemMod) -> syn::Result<Self> {
        let pvq_extension = generate_crate_access("pvq-extension")?;
        let item_span = item.span();
        let mod_name = &item.ident;
        let items = &mut item
            .content
            .as_mut()
            .ok_or_else(|| {
                let msg = "Invalid extension_decl definition, expected mod to be inline.";
                syn::Error::new(item_span, msg)
            })?
            .1;
        let mut extension = None;
        for item in items.iter_mut() {
            let extension_attr: Option<ExtensionDeclAttr> = helper::take_first_item_extension_decl_attr(item)?;

            match extension_attr {
                Some(ExtensionDeclAttr::Extension(_)) if extension.is_none() => {
                    extension = Some(extension::Extension::try_from(&mod_name.to_string(), item)?);
                }
                Some(attr) => {
                    let msg = "Invalid duplicated attribute";
                    return Err(syn::Error::new(attr.span(), msg));
                }
                None => (),
            }
        }

        Ok(Self {
            item,
            extension: extension.ok_or_else(|| syn::Error::new(item_span, "Missing `#[extension_decl::extension]`"))?,
            pvq_extension,
        })
    }
}

/// Parse attributes for item in extension module
/// syntax must be `extension_decl::` (e.g. `#[extension_decl::config]`)
enum ExtensionDeclAttr {
    Extension(proc_macro2::Span),
}

impl syn::parse::Parse for ExtensionDeclAttr {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        input.parse::<syn::Token![#]>()?;
        let content;
        syn::bracketed!(content in input);
        content.parse::<keyword::extension_decl>()?;
        content.parse::<syn::Token![::]>()?;

        let lookahead = content.lookahead1();
        if lookahead.peek(keyword::extension) {
            let span = content.parse::<keyword::extension>()?.span();
            Ok(Self::Extension(span))
        } else {
            Err(lookahead.error())
        }
    }
}

impl ExtensionDeclAttr {
    fn span(&self) -> proc_macro2::Span {
        match self {
            Self::Extension(span) => *span,
        }
    }
}
