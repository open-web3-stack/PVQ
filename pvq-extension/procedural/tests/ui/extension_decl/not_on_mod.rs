use pvq_extension_procedural::extension_decl;

// This should fail because extension_decl can only be used on modules
#[extension_decl]
struct InvalidUsage {
    field: u32,
}

fn main() {}
