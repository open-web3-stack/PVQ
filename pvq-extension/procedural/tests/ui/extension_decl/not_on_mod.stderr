error: expected `mod`
 --> tests/ui/extension_decl/not_on_mod.rs:5:1
  |
5 | struct InvalidUsage {
  | ^^^^^^
